### PR TITLE
[Hetero] Support heterogeneous training strategy with different `dp_size`

### PR DESCRIPTION
**Pull request by heavyrain-lzy**
Support heterogeneous training strategy with different `dp_size`.

### CODE REVIEW

File: flagscale/train/hetero/parallel_context.py [hunk 1]
```diff
@@ -532,16 +532,40 @@ def build_inter_mesh_process_groups(self, process_mesh1, process_mesh2):
...
+                        dst_dp_dims = [dim for dim, _, _ in dp_overlapped_mapping[k]]
+                        dst_coords = list(
+                            itertools.product([0], [0], dst_dp_dims, dst_pp_dims)
+                        )
+                        src_rank = process_mesh1.logical_coords_to_physical_ranks(
+                            [src_coord]
+                        )[0]
+                        for dst_coord in dst_coords:
+                            tp_dim, cp_dim, dp_dim, pp_dim = dst_coord
+                            dst_coord = [tp_dim, cp_dim, dp_dim, pp_dim]
```
Comment (aoyulong) [inline flagscale/train/hetero/parallel_context.py:563]: Are these two statements duplicated?

**Review by aoyulong — APPROVED** [review]
LGTM

File: flagscale/train/hetero/parallel_context.py (hunk 1 above)
Comment (heavyrain-lzy) [reply flagscale/train/hetero/parallel_context.py:563]: Thanks. I'll fix it next PR.

**Review by aoyulong — APPROVED** [review]
LGTM

**Pull request merged**